MF-532: Implement an E2E test for the login flow of refapp-3.x

jayasanka-sack opened the pull request:
## Purpose
The purpose of this pr is to fix [MF-532](https://issues.openmrs.org/browse/MF-532)

## Goals
To give users confidence that the Login workflow of the RefApp 3.x is working as expected.

## Approach
- Create a new property file to store properties
- Remove step definition files of the ocl client (Because it has been already moved to a separate project)
- Implement an E2E test for the login flow of refapp-3.x
- Create a GitHub workflow file
- Increase the max waiting time to 40 seconds (Sometimes the test server takes that much time to load the inner components but it pretends the page has been fully loaded. Therefore increasing the waiting time is essential)
- Add workflow badge to the readme file

## Demo
![Kapture 2021-07-12 at 22 59 28](https://user-images.githubusercontent.com/33048395/125331041-18128100-e365-11eb-9e1d-b3a15bc47a19.gif)

jayasanka-sack commented on src/test/resources/org/openmrs/uitestframework/test-cypress.properties:
```diff
@@ -0,0 +1 @@
+webapp.url=https://openmrs-spa.org/openmrs/spa
```
Is this the baseurl that are we going to use? 

Will it be suitable for the test that has "create" operations? (ex: patient registration)

kaweesi replied:
yeah, that should be the base-url

brandones replied:
I thought that we were going to be running tests on isolated test servers.

kaweesi replied:
a test instance for microfrontends should be started in the github actions and waited for before running tests whose url should used here insyead

brandones replied:
Great. I think that @jayasanka-sack will need some help making that happen, and that we probably shouldn't block this PR on it.

jayasanka-sack commented:
@kaweesi Could you please review this? 😇

kaweesi commented on cypress/integration/cucumber/step_definitions/refapp-3.x/01-login/login.js:
```diff
@@ -0,0 +1,24 @@
+import {Given} from 'cypress-cucumber-preprocessor/steps';
+
+Given('user arrives at the login page', () => {
+    cy.task('getProperty', 'webapp.url').then(baseUrl => {
```
am wondering if we should extract out a re-usable function/command for visiting a url. otherwise this looks nice

jayasanka-sack replied:
Yeah that would be nice! I'll do it!

jayasanka-sack replied:
Fixed.

kaweesi commented on cypress/integration/cucumber/step_definitions/refapp-3.x/01-login/login.js:
```diff
@@ -0,0 +1,24 @@
+    cy.contains('Continue').click();
+    cy.get('#password').type(password);
+    cy.contains('Log in').click();
+    cy.wait(3000);
```
any reason why you want to delay here?

jayasanka-sack replied:
Actually, it takes a long time to load. But later on, I figured out a way to increase the waiting time (by updating the cypress json file). I'll update the PR

jayasanka-sack replied:
fixed.

jayasanka-sack commented:
@kaweesi, @brandones  I changed the structure of the workflow (inspired by a previous implementation on ref app 2.x) 
Also added a GitHub workflow as well. Could you please review this when you are not busy?

brandones commented on cypress/integration/cucumber/step_definitions/refapp-3.x/01-login/login.js:
```diff
@@ -0,0 +1,35 @@
+    cy.contains('Log in').click();
+    // The user will be redirected to the location selection page if the credentials are correct.
+    // The following artificial wait is to wait until it changes the url.
+    cy.wait(2000);
```
Don't use static `wait`s. Please see [the Cypress documentation](https://docs.cypress.io/faq/questions/using-cypress-faq#How-can-I-make-Cypress-wait-until-something-is-visible-in-the-DOM).

brandones commented on cypress/integration/cucumber/step_definitions/refapp-3.x/01-login/login.js:
```diff
@@ -0,0 +1,35 @@
+})
+
+When('the user logs in with {string} and {string} to the {string}', (username, password, location) => {
+    cy.get('#username').type(username);
```
Don't get elements by ID. IDs are implementation details. Get elements by how they appear to the user. In this case, it is a text input with label `Username`.

kaweesi replied:
do you have a plan to maintain the ids permanent? if so then i would strongly think ids/classes etc are a great ways to refer to elements vs labels which may change with locale changes etc which could simply be out of the test scope

brandones replied:
No; ids and classes are implementation details. "Implementation details" are aspects of the code which are invisible to the user. [Do not test implementation details](https://kentcdodds.com/blog/testing-implementation-details#why-is-testing-implementation-details-bad). We should test against the English labels. Those are the "source" labels in the code. If those change, it will be highly apparent, and easy to adapt the tests. AFAIK locale changes are not currently in test scope; but testing them is not out of the question.

brandones replied:
This is why I have recommended we use [Cypress Testing Library](https://testing-library.com/docs/cypress-testing-library/intro/)—it provides very natural semantics for "testing from the user's perspective." See also the [Testing Library Introduction](https://testing-library.com/docs/). This video, "[Build invincible tests](https://www.youtube.com/watch?v=F8LH9d9eN3M)," is also very good.
